Modelica/Resources/C-Sources/ModelicaTime.c

```diff
@@ -0,0 +1,260 @@
+    tlocal.tm_isdst = -1;
+    tlocal.tm_wday = 0;
+    tlocal.tm_yday = 0;
+
+    calendarTime = mktime(&tlocal);
+    if (-1 == calendarTime) {
+        ModelicaFormatError("Not possible to convert \"%4i-%02i-%02i %02i:%02i:%02i\" "
+            "to time_t type in local time zone.\n", year, mon, mday, hour, min, sec);
+    }
+    return calendarTime;
+}
+#endif
+
+double ModelicaTime_difftime(int ms, int sec, int min, int hour, int mday,
+                             int mon, int year, int refYear) {
+    /* Get elapsed seconds w.r.t. reference year */
+#if defined(NO_TIME)
+    return 0.0;
+#else
+    if (1 == mday && 1 == mon) {
+        if (year == refYear) {
+            return 60 * (60 * hour + min) + sec + ms/1000.0;
+        }
+        else if (1970 == year) {
+            /* Avoid calling mktime for New Year 1970 in local time zone */
+            const time_t endTime = epoch(sec, min, hour, 2, 1, 1970);
+            const time_t startTime = epoch(0, 0, 0, 1, 1, refYear);
+            return difftime(endTime, startTime) - 86400.0 + ms/1000.0;
+        }
+    }
+    {
+        const time_t endTime = epoch(sec, min, hour, mday, mon, year);
+        if (1970 == refYear) {
+            /* Avoid calling mktime for New Year 1970 in local time zone */
+            const time_t startTime = epoch(0, 0, 0, 2, 1, 1970);
+            return difftime(endTime, startTime) + 86400.0 + ms/1000.0;
+        }
+        else {
+            const time_t startTime = epoch(0, 0, 0, 1, 1, refYear);
+            return difftime(endTime, startTime) + ms/1000.0;
+        }
+    }
+#endif
+}
+
+void ModelicaTime_localtime(_Out_ int* ms, _Out_ int* sec, _Out_ int* min,
+                            _Out_ int* hour, _Out_ int* mday, _Out_ int* mon,
+                            _Out_ int* year, double seconds, int refYear) {
+    /* Convert elapsed seconds w.r.t. reference year */
+#if defined(NO_TIME)
+    *ms   = 0;
+    *sec  = 0;
+    *min  = 0;
+    *hour = 0;
+    *mday = 0;
+    *mon  = 0;
+    *year = 0;
+#else
+    struct tm* tlocal;
+    time_t calendarTime;
+    double intPartOfSeconds;
+    double fracPartOfSeconds;
+#if defined(_POSIX_) || (defined(_MSC_VER) && _MSC_VER >= 1400)
+    struct tm tres;
+#endif
+
+    fracPartOfSeconds = modf(seconds, &intPartOfSeconds);
+    calendarTime = (time_t)intPartOfSeconds;
+
+    if (1970 == refYear) {
+        /* Avoid calling mktime for New Year 1970 in local time zone */
+        calendarTime = calendarTime + epoch(0, 0, 0, 2, 1, 1970) - (time_t)86400;
+    }
+    else {
+        calendarTime = calendarTime + epoch(0, 0, 0, 1, 1, refYear);
+    }
+
+#if defined(_POSIX_)
+    tlocal = localtime_r(&calendarTime, &tres); /* Time fields in local time zone */
+#elif defined(_MSC_VER) && _MSC_VER >= 1400
+    localtime_s(&tres, &calendarTime);          /* Time fields in local time zone */
+    tlocal = &tres;
+#else
+    tlocal = localtime(&calendarTime);          /* Time fields in local time zone */
```

Comment:
I would have preferred if we used tres after this and only in this case used tlocal at all, and use tres=*tlocal; after the call of localtime as copying a struct has been in C since the early days (it might have been missing in some variants prior to C89/C90). (And time.h should, since C89/C90 define 'struct tm' in time.h.)

Reply:
I see. FYI, the above implementation matches the current implementation of MSL 3.2.3 or 4.0.0

https://github.com/modelica/ModelicaStandardLibrary/blob/1adccd9a281c14d89abec1155a41c486984c1d74/Modelica/Resources/C-Sources/ModelicaInternal.c#L1297-L1304

Reply:
Ok, I still think the change would be good, but it's not that important.